### PR TITLE
RFC-82: Implement support for Entity Tags

### LINKED ISSUE #47: Add support for RFC 82 (entity tags)
[RFC 82](https://github.com/cedar-policy/rfcs/blob/emina/entity-tags/text/0082-entity-tags.md) was recently accepted and is supported in the Rust library as of version [4.2.0](https://github.com/cedar-policy/cedar/releases/tag/v4.2.0).

### LINKED ISSUE #52: Upstream Integration Test Corpus Modified
The upstream integration test corpus at https://raw.githubusercontent.com/cedar-policy/cedar-integration-tests/main/corpus-tests.tar.gz has been updated. Please integrate the changes into the local copy.

### PR DESCRIPTION
**Pull request by adowair**
This pull request will implement support for Cedar Entity Tags per [RFC-82](https://github.com/cedar-policy/rfcs/blob/7fcdbb509dd2709e03897fe40accbefd8fdb56f3/text/0082-entity-tags.md).

This pull request will also update the [corpus tests](https://github.com/cedar-policy/cedar-integration-tests) to the latest version: https://github.com/cedar-policy/cedar-integration-tests/pull/12

Closes #47
Closes #52

#### Test Coverage
```console
$ git checkout main
$ go test -count 1 -cover ./... | grep -v -E 'coverage: 100|no test files'
ok      github.com/cedar-policy/cedar-go/ast    0.257scoverage: 98.1% of statements
$ git checkout tag-support
$ go test -count 1 -cover ./... | grep -v -E 'coverage: 100|no test files'
ok      github.com/cedar-policy/cedar-go/ast    1.315scoverage: 98.2% of statements
```

### CODE REVIEW

File: internal/eval/fold_test.go [hunk 1]
```diff
@@ -473,6 +473,26 @@ func TestFoldPolicy(t *testing.T) {
...
+		},
+		{
+			"opGetTagEntity",
+			ast.Permit().When(ast.EntityUID("T", "1").GetTag(ast.String("key"))),
+			ast.Permit().When(ast.EntityUID("T", "1").GetTag(ast.String("key"))),
+		},
+		{
+			"opHasTagEntity",
+			ast.Permit().When(ast.EntityUID("T", "1").HasTag(ast.String("key"))),
+			ast.Permit().When(ast.False()),
```
Comment (adowair) [inline internal/eval/fold_test.go:494]: @philhassey @patjakdev notice that `hasTag()` nodes are _folded_ when there are no entities in the context. This is because `<non_existing_medallion>.hasTag("someTag")` evaluates to `false`.

File: authorize.go [hunk 2]
```diff
@@ -19,6 +19,9 @@ const (
 // IsAuthorized uses the combination of the PolicySet and Entities to determine
 // if the given Request to determine Decision and Diagnostic.
 func (p PolicySet) IsAuthorized(entities types.EntityGetter, req Request) (Decision, Diagnostic) {
+	if entities == nil {
+		entities = types.EntityMap{}
+	}
```
Comment (patjakdev) [inline authorize.go:24]: Oh Golang...

File: authorize_test.go [hunk 3]
```diff
@@ -784,7 +795,7 @@ func TestIsAuthorized(t *testing.T) {
 	for _, tt := range tests {
 		tt := tt
 		t.Run(tt.Name, func(t *testing.T) {
-			t.Parallel()
+			// t.Parallel()
```
Comment (patjakdev) [inline authorize_test.go:798]: Please add this back.

File: authorize_test.go [hunk 4]
```diff
@@ -33,6 +34,16 @@ func TestIsAuthorized(t *testing.T) {
...
+		{
+			Name:      "nil-entity-map",
+			Policy:    `permit(principal,action,resource);`,
+			Principal: cuzco,
+			Action:    dropTable,
+			Resource:  cedar.NewEntityUID("table", "whatever"),
+			Context:   cedar.Record{},
+			Want:      true,
+			DiagErr:   0,
+		},
```
Comment (patjakdev) [inline authorize_test.go:46]: Can you explicitly initialize `Entities` to `nil` here?

Also, I guess I would name the test `nil-entity-getter`.

File: corpus_test.go [hunk 5]
```diff
@@ -244,11 +245,27 @@ func TestCorpusRelated(t *testing.T) {
...
+			resource
+			) when {
+			true && (resource.hasTag("A"))
+			};`,
+			types.EntityMap{cedar.NewEntityUID("a", ""): cedar.Entity{Attributes: cedar.NewRecord(cedar.RecordMap{"A": types.False})}},
+			cedar.Request{Principal: cedar.NewEntityUID("a", ""), Action: cedar.NewEntityUID("Action", "action"), Resource: cedar.NewEntityUID("a", "'")},
+			cedar.Deny,
+			nil,
+			nil,
+		},
```
Comment (patjakdev) [inline corpus_test.go:268]: I'm not 100% convinced we should check this in. I think it would make sense instead to add a pared down version of this to `TestIsAuthorized()`.

For example, I don't think we actually care about the duplicate `Action::"action"` in the action set. You want to validate some very specific behaviors about `getTag()` and `hasTag()`.

File: internal/eval/evalers.go [hunk 6]
```diff
@@ -804,6 +805,73 @@ func (n *hasEval) Eval(env Env) (types.Value, error) {
...
+func (n *getTagEval) Eval(env Env) (types.Value, error) {
+	eid, err := evalEntity(n.lhs, env)
+	if err != nil {
+		return zeroValue(), err
+	}
+
+	var unspecified types.EntityUID
+	if eid == unspecified {
+		return zeroValue(), fmt.Errorf("cannot access tag `%s` of %w", n.rhs, errUnspecifiedEntity)
+	}
```
Comment (patjakdev) [inline internal/eval/evalers.go:826]: I think I would call this a "zero" ID rather than an "unspecified" ID.

That being said, it appears that entity type names cannot be empty, but the id part of the string can be. The Cedar text unmarshaler wouldn't allow this entity to exist, but I think the JSON unmarshaler and `NewEntityUID()` would.

I think we should actually file a bug on the JSON unmarshaler for allowing empty type names. We should also really enforce the type name being non-empty in `NewEntityUID()`, but adding an error return type to that function would be a breaking change. Maybe we could panic?

All that is to say, I'm not sure how important it really is to consider this case in the evaler.

File: internal/eval/evalers.go [hunk 7]
```diff
@@ -1139,9 +1207,13 @@ func newExtensionEval(name types.Path, args []Evaler) Evaler {
...
 	if i, ok := extensions.ExtMap[name]; ok {
 		if i.Args != len(args) {
-			return newErrorEval(fmt.Errorf("%w: %s takes %d parameter(s)", errArity, name, i.Args))
+			return newErrorEval(fmt.Errorf("%w: %s takes %d parameter(s), but %d provided", errArity, name, i.Args, len(args)))
 		}
 		switch {
+		case name == "hasTag":
+			return newHasTagEval(args[0], args[1])
+		case name == "getTag":
+			return newGetTagEval(args[0], args[1])
```
Comment (patjakdev) [inline internal/eval/evalers.go:1216]: `hasTag` and `getTag` are not extension methods. I would think they should be implemented as first class operators, like `contains()` and its ilk.

File: internal/eval/evalers_test.go [hunk 8]
```diff
@@ -1448,6 +1448,175 @@ func TestHasNode(t *testing.T) {
...
+			newLiteralEval(unknownTag),
+			zeroValue(),
+			errTagAccess},
+		{"UnknownEntity",
+			newLiteralEval(types.NewEntityUID("unknownType", unknownID)),
+			newLiteralEval(knownTag),
+			zeroValue(),
+			errEntityNotExist},
+		{"UnspecifiedEntity",
+			newLiteralEval(types.NewEntityUID("", "")),
```
Comment (patjakdev) [inline internal/eval/evalers_test.go:1509]: Yeah, this totally shouldn't be allowed.

File: internal/eval/fold_test.go (hunk 1 above)
Comment (patjakdev) [reply internal/eval/fold_test.go:494]: Isn't this...wrong? Maybe I'm missing something, but we can't be allowed to fold this to false because during folding we don't know which entities actually exist.

The `has` operator works like I expect. Even though it returns false if the entity doesn't exist, it doesn't fold away here...(see `opHasEntity` above).

File: internal/extensions/extensions.go [hunk 9]
```diff
@@ -8,6 +8,9 @@ type extInfo struct {
 }
 
 var ExtMap = map[types.Path]extInfo{
+	"hasTag": {Args: 2, IsMethod: true},
+	"getTag": {Args: 2, IsMethod: true},
+
```
Comment (patjakdev) [inline internal/extensions/extensions.go:13]: As I mentioned above, these methods should be treated as  first-class operators, not extension methods.

File: types/entity.go [hunk 10]
```diff
@@ -11,6 +11,7 @@ type Entity struct {
 	UID        EntityUID    `json:"uid"`
 	Parents    EntityUIDSet `json:"parents"`
 	Attributes Record       `json:"attrs"`
+	Tags       Record       `json:"tags,omitempty"`
```
Comment (patjakdev) [inline types/entity.go:14]: You determined that `omitempty` is useless here, right?

File: types/entity_tag_test.go [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+import (
+	"testing"
+	"time"
+
+	"github.com/cedar-policy/cedar-go"
+	"github.com/cedar-policy/cedar-go/internal/testutil"
+	"github.com/cedar-policy/cedar-go/types"
+)
+
+func TestEntityTagMarshalJSON(t *testing.T) {
```
Comment (patjakdev) [inline types/entity_tag_test.go:12]: I would put this in `types.entity_test.go`. I don't see a reason for it to live in its own file, really.

**Review by patjakdev — CHANGES_REQUESTED** [review]

File: authorize.go (hunk 2 above)
Comment (philhassey) [reply authorize.go:24]: Lol, I think ... slightly better might be (again, this is Go):

```
var zero types.EntityMap
entities = zero
```

Which is because (and feel free to Go playground / check on this explanation) when you just say `var zero types.EntityMap` you are yielding what amounts to a typed nil pointer.  However, when you do `entities = types.EntityMap{}` you are allocating an empty map.

In our case, probably either would work fine, but the `var zero` more precisely replicates the original behavior and avoids an unnecessary allocation.

File: internal/extensions/extensions.go (hunk 9 above)
Comment (adowair) [reply internal/extensions/extensions.go]: Thank you, done here: https://github.com/cedar-policy/cedar-go/pull/77/commits/4f433e34c428f2de144c83035ba1fd0a313a904c

File: internal/eval/evalers.go (hunk 7 above)
Comment (adowair) [reply internal/eval/evalers.go]: Yes, done here: https://github.com/cedar-policy/cedar-go/pull/77/commits/4f433e34c428f2de144c83035ba1fd0a313a904c.

File: types/entity_tag_test.go (hunk 11 above)
Comment (adowair) [reply types/entity_tag_test.go]: Yes, done here: https://github.com/cedar-policy/cedar-go/pull/77/commits/9074d5b00f44be4346c13ac8b7270a09b06c541a.

File: types/entity.go (hunk 10 above)
Comment (adowair) [reply types/entity.go]: Yes, removed here: https://github.com/cedar-policy/cedar-go/pull/77/commits/9074d5b00f44be4346c13ac8b7270a09b06c541a.

File: internal/eval/evalers_test.go (hunk 8 above)
Comment (adowair) [reply internal/eval/evalers_test.go:1509]: I've renamed this to `ZeroEntity`—while it's not a possible call path, we do need this test to maintain 100% net new code coverage. Done here: https://github.com/cedar-policy/cedar-go/pull/77/commits/9074d5b00f44be4346c13ac8b7270a09b06c541a.

File: internal/eval/evalers.go (hunk 6 above)
Comment (adowair) [reply internal/eval/evalers.go]: Yes, I think that makes sense, thanks for expounding on it!

I've renamed this to `zero` here: https://github.com/cedar-policy/cedar-go/pull/77/commits/9074d5b00f44be4346c13ac8b7270a09b06c541a.

File: corpus_test.go (hunk 5 above)
Comment (adowair) [reply corpus_test.go]: Yes, done here: https://github.com/cedar-policy/cedar-go/pull/77/commits/9074d5b00f44be4346c13ac8b7270a09b06c541a.

File: authorize_test.go (hunk 4 above)
Comment (adowair) [reply authorize_test.go]: Yes, done here: https://github.com/cedar-policy/cedar-go/pull/77/commits/9074d5b00f44be4346c13ac8b7270a09b06c541a.

File: authorize_test.go (hunk 3 above)
Comment (adowair) [reply authorize_test.go]: Yes, done here: https://github.com/cedar-policy/cedar-go/pull/77/commits/9074d5b00f44be4346c13ac8b7270a09b06c541a.

File: ast/ast_test.go [hunk 12]
```diff
@@ -347,6 +347,16 @@ func TestASTByTable(t *testing.T) {
 			ast.Permit().When(ast.Long(42).Has("key")),
 			internalast.Permit().When(internalast.Long(42).Has("key")),
 		},
+		{
+			"opGetTag",
+			ast.Permit().When(ast.Long(42).GetTag(ast.String("key"))),
```
Comment (patjakdev) [inline ast/ast_test.go:352]: nit: I think we should strive for these to be "valid" ASTs, so maybe change those `Long`s to `EntityUID`s.

File: authorize.go (hunk 2 above)
Comment (patjakdev) [reply authorize.go:24]: Agreed with Phil on avoiding the extra allocation, as nitpicky as it may be.

I'm not sure who would ever pass a nil `EntityGetter`.

File: internal/eval/fold_test.go [hunk 13]
```diff
@@ -473,6 +473,26 @@ func TestFoldPolicy(t *testing.T) {
...
+		{
+			"opHasTag",
+			ast.Permit().When(ast.Long(42).HasTag(ast.String("key"))),
+			ast.Permit().When(ast.Long(42).HasTag(ast.String("key"))),
+		},
+		{
+			"opGetTagEntity",
+			ast.Permit().When(ast.EntityUID("T", "1").GetTag(ast.String("key"))),
+			ast.Permit().When(ast.EntityUID("T", "1").GetTag(ast.String("key"))),
+		},
```
Comment (patjakdev) [inline internal/eval/fold_test.go:490]: It feels like this ought to be the `opGetTag` test and the current `opGetTag` test should be `opGetTagInvalidType` test or the like.

File: internal/eval/partial_test.go [hunk 14]
```diff
@@ -1258,6 +1282,63 @@ func TestPartialBasic(t *testing.T) {
 	}
 }
 
+func TestPartialWithContext(t *testing.T) {
+	t.Parallel()
+	tests := []struct {
+		name       string
+		context    types.Value
+		entitities types.EntityGetter
```
Comment (patjakdev) [inline internal/eval/partial_test.go:1290]: ahem

File: internal/parser/cedar_marshal.go [hunk 15]
```diff
@@ -364,6 +380,8 @@ func astNodeToMarshalNode(astNode ast.IsNode) IsNode {
 		return NodeTypeNot{v, UnaryNode{}}
 	case ast.NodeTypeAccess:
 		return NodeTypeAccess{v}
+	case ast.NodeTypeGetTag:
+		return NodeTypeGetTag{v}
```
Comment (patjakdev) [inline internal/parser/cedar_marshal.go:384]: Did the two cases get separated by accident? It doesn't seem to be alphabetical or some other order I can discern.

File: x/exp/ast/node.go [hunk 16]
```diff
@@ -67,6 +67,10 @@ type NodeTypeHas struct {
 	StrOpNode
 }
 
+type NodeTypeHasTag struct {
```
Comment (patjakdev) [inline x/exp/ast/node.go:70]: Keep it one one line?

File: x/exp/batch/batch.go [hunk 17]
```diff
@@ -155,6 +155,9 @@ func Authorize(ctx context.Context, ps *cedar.PolicySet, entities types.EntityGe
 	case request.Context == nil:
 		return fmt.Errorf("%w: context", errMissingPart)
 	}
+	if entities == nil {
```
Comment (patjakdev) [inline x/exp/batch/batch.go:158]: Same comment about avoiding allocation

**Review by patjakdev — APPROVED** [review]
Nice!!! Just a couple little nitpicks.

File: internal/eval/partial_test.go (hunk 14 above)
Comment (adowair) [reply internal/eval/partial_test.go:1290]: oh my goodness

File: internal/eval/fold.go [hunk 18]
```diff
@@ -109,6 +109,32 @@ func fold(n ast.IsNode) ast.IsNode {
 				return ast.NodeTypeHas{StrOpNode: ast.StrOpNode{Arg: nodes[0], Value: v.Value}}
 			},
 		)
+	case ast.NodeTypeGetTag:
+		return tryFold(
+			[]ast.IsNode{v.Left, v.Right},
+			func(values []types.Value) Evaler {
+				if _, ok := values[0].(types.EntityUID); ok {
```
Comment (philhassey) [inline internal/eval/fold.go:116]: You could probably just return an error always for both this and HasTag

```
func(values []types.Value) Evaler {
    return newErrorEval(fmt.Errorf("fold.GetTag.EntityUID"))
}
```

File: types/entity.go [hunk 19]
```diff
@@ -33,10 +34,12 @@ func (e Entity) MarshalJSON() ([]byte, error) {
 		UID        ImplicitlyMarshaledEntityUID   `json:"uid"`
 		Parents    []ImplicitlyMarshaledEntityUID `json:"parents"`
 		Attributes Record                         `json:"attrs"`
+		Tags       Record                         `json:"tags"`
```
Comment (philhassey) [inline types/entity.go:37]: @patjakdev did you mention that for Rust parity this should be `omitempty` ?

Comment (philhassey) [reply types/entity.go:37]: I guess, either way, a test showing it does what we want it to do for the empty tags case would be nice.

Comment (philhassey) [reply types/entity.go:37]: (I see a test exists for this, so that's nice.)

**Review by philhassey — APPROVED** [review]
LGTM - just a few tiny nits left

**Comment by philhassey** [comment]
There is a lint error to address as well:

image

File: internal/eval/fold_test.go (hunk 1 above)
Comment (adowair) [reply internal/eval/fold_test.go]: Yes, correct on all accounts! Fixed this up here: https://github.com/cedar-policy/cedar-go/pull/77/commits/4f433e34c428f2de144c83035ba1fd0a313a904c.

File: types/entity.go (hunk 19 above)
Comment (adowair) [reply types/entity.go:37]: Yes, we deliberated on this here: https://github.com/strongdm/cedar-go/pull/11#discussion_r1870037794

File: internal/parser/cedar_marshal.go (hunk 15 above)
Comment (adowair) [reply internal/parser/cedar_marshal.go:384]: I've separated them to each respectively be right after the equivalent attribute operation.

File: internal/eval/fold.go (hunk 18 above)
Comment (adowair) [reply internal/eval/fold.go]: Yes, done here: https://github.com/cedar-policy/cedar-go/pull/77/commits/36fd3e7573bac8103b345a0c8d413d6eef37dccb.

File: x/exp/batch/batch.go (hunk 17 above)
Comment (adowair) [reply x/exp/batch/batch.go:158]: Yes, done here: https://github.com/cedar-policy/cedar-go/pull/77/commits/36fd3e7573bac8103b345a0c8d413d6eef37dccb.

File: x/exp/ast/node.go (hunk 16 above)
Comment (adowair) [reply x/exp/ast/node.go]: Yes, done here: https://github.com/cedar-policy/cedar-go/pull/77/commits/36fd3e7573bac8103b345a0c8d413d6eef37dccb.

File: internal/eval/partial_test.go (hunk 14 above)
Comment (adowair) [reply internal/eval/partial_test.go]: Fixed here: https://github.com/cedar-policy/cedar-go/pull/77/commits/36fd3e7573bac8103b345a0c8d413d6eef37dccb.

File: internal/eval/fold_test.go (hunk 13 above)
Comment (adowair) [reply internal/eval/fold_test.go]: Yes, done here: https://github.com/cedar-policy/cedar-go/pull/77/commits/36fd3e7573bac8103b345a0c8d413d6eef37dccb.

File: ast/ast_test.go (hunk 12 above)
Comment (adowair) [reply ast/ast_test.go]: Yes, done here: https://github.com/cedar-policy/cedar-go/pull/77/commits/36fd3e7573bac8103b345a0c8d413d6eef37dccb.

**Comment by adowair** [comment]
Notice that this is a slightly breaking change w.r.t. the backward compatibility of JSON marshalling. Now `tags: {}` will be emitted as a part of entity struct, since empty structs cannot be omitted by the `omitempty` tag. Go `v1.24` will introduce an `omitzero` tag which solves this exact problem.
